#**Maestría en Inteligencia Artificial Aplicada**
##**Curso: Procesamiento de Lenguaje Natural (NLP)**
###Tecnológico de Monterrey
###Prof Luis Eduardo Falcón Morales

## **Adtividad de la Semana 02**
###**Introducción al procesamiento de texto.**

En esta actividad deberás utilizar los datos del siguiente archivo que se encuentra en Canvas:

MNA_NLP_semana_02_Actividad_datos.txt

El archivo contiene comentarios en inglés sobre servicios de comida de la página de Yelp: https://www.yelp.com/ . 

Son mil comentarios y forman parte del conjunto de datos que se encuentra en el Machine Learning Repository de la UCI, llamado "Sentiment Labelled Sentences": https://archive.ics.uci.edu/ml/datasets/Sentiment+Labelled+Sentences#


#**Parte 1. Cargamos los datos.**   

Cargar los datos del archivo indicado y obtener una lista de longitud de 1000 strings/comentarios. 

Por el momento solamente requerimos las bibliotecas de Numpy y re, para el manejo de los arreglos y de las expresiones regulares en Python.

En particular, no necesitarás en esta actividad la biblioteca de Pandas.

###**NOTA: En esta actividad no debes importar nada más, con estas dos bibliotecas será *suficiente*.**

In [39]:
import numpy as np    # importamos Numpy para el manejo de los arreglos.
import re             # importamos re para el manejo de las expresiones regulares.

In [40]:
# Ejecuta las siguientes instrucciones para cargar la información del achivo dado:

with open('MNA_NLP_semana_02_Actividad_datos.txt',        # puedes actualizar la ruta a tu archivo, en dado caso.
          mode='r',     # abrimos el archivo en modo lectura.
          ) as f:
    docs = f.readlines()    # separamos cada comentario por líneas

f.close()  # ya que tenemos la información en la variable docs, cerramos el archivo

In [41]:
type(docs) == list   # Verifica que tu variable "docs" es una lista

True

In [42]:
len(docs)==1000  # verifica que la longitud de "docs" es de mil comentarios.

True

In [43]:
docs[0:10]     # observa algunos de los primeros comentarios

['Wow... Loved this place.\n',
 'Crust is not good.\n',
 'Not tasty and the texture was just nasty.\n',
 'Stopped by during the late May bank holiday off Rick Steve recommendation and loved it.\n',
 'The selection on the menu was great and so were the prices.\n',
 'Now I am getting angry and I want my damn pho.\n',
 "Honeslty it didn't taste THAT fresh.)\n",
 'The potatoes were like rubber and you could tell they had been made up ahead of time being kept under a warmer.\n',
 'The fries were great too.\n',
 'A great touch.\n']

#**Parte 2: sección de preguntas (regex).**   


##**Instrucciones:**

###**A continuación deberás contestar cada una de las preguntas que te piden usando expresiones regulares (regex).**

###**Por el momento no hay restricción en cuanto al número de líneas de código que agregues, pero trata de incluir las mínimas posibles.**

*   **Pregunta 1.** 

Busca y elimina todos los saltos de línea '\n' que se encuentran al final de cada comentario. 

Una vez finalizado, imprime los primeros 10 comentarios del resultado obtenido.


In [44]:
# Usamos una lista de comprensión para aplicar re.sub a cada comentario (c) en la lista docs.
# El patrón r'\n$' busca el salto de línea específicamente al final de la cadena.

com_clean = [re.sub(r'\n$', '', c) for c in docs]

In [45]:
# Imprimimos los primeros 10 elementos de la lista ya limpia para verificar el cambio.
print(f"Comentarios sin salto de línea: {com_clean[:10]}")

# Mostramos el conteo total de elementos para asegurar que seguimos teniendo los 1000 originales.
print(f"Total de comentarios sin el salto de línea: {len(com_clean)}")

Comentarios sin salto de línea: ['Wow... Loved this place.', 'Crust is not good.', 'Not tasty and the texture was just nasty.', 'Stopped by during the late May bank holiday off Rick Steve recommendation and loved it.', 'The selection on the menu was great and so were the prices.', 'Now I am getting angry and I want my damn pho.', "Honeslty it didn't taste THAT fresh.)", 'The potatoes were like rubber and you could tell they had been made up ahead of time being kept under a warmer.', 'The fries were great too.', 'A great touch.']
Total de comentarios sin el salto de línea: 1000


*   **Pregunta 2.**  

Busca e imprime todas las palabras que terminan con dos o más signos de admiración seguidos, por ejemplo "!!!". 

Debes imprimir tanto la palabra como la totalidad de signos de admiración que le siguen. 

Indica cuántos resultados obtuviste.



In [46]:
# Unimos todos los comentarios en un solo string y buscamos palabras (\w+) 
# que terminen con 2 o más signos de admiración (!{2,}) de forma consecutiva.

sign_adm = re.findall(r'\w+!{2,}', " ".join(docs))

In [47]:
# Imprimimos el conteo total de palabras que cumplen con el patrón de énfasis.
print(f"Total de palabras encontradas: {len(sign_adm)}")

# Mostramos las palabras encontradas.
print(f"Palabras encontradas: {sign_adm}")


Total de palabras encontradas: 26
Palabras encontradas: ['Firehouse!!!!!', 'APPETIZERS!!!', 'amazing!!!', 'buffet!!!', 'good!!', 'it!!!!', 'DELICIOUS!!', 'amazing!!', 'shawarrrrrrma!!!!!!', 'yucky!!!', 'steak!!!!!', 'delicious!!!', 'far!!', 'biscuits!!!', 'dry!!', 'disappointing!!!', 'awesome!!', 'Up!!', 'FLY!!!!!!!!', 'here!!!', 'great!!!!!!!!!!!!!!', 'packed!!', 'otherwise!!', 'amazing!!!!!!!!!!!!!!!!!!!', 'style!!', 'disappointed!!']


*   **Pregunta 3.**  

Busca e imprime todas las palabras que están escritas totalmente en mayúsculas. Cada coincidencia debe ser una sola palabra.

Indica cuántas palabras encontraste.



In [48]:
# Utilizamos findall para buscar todas las ocurrencias que coincidan con el patrón.
# \b marca el límite de una palabra, y [A-ZÁÉÍÓÚÑ]+ busca uno o más caracteres en mayúscula, incluyendo acentos.

mayus_words = re.findall(r'\b[A-ZÁÉÍÓÚÑ]+\b', " ".join(docs))

In [49]:
# Mostramos la cantidad total de palabras encontradas que cumplen con ser totalmente mayúsculas.
print(f"Total de palabras encontradas: {len(mayus_words)}")

# Imprimimos todos los elementos de la lista para verificar el tipo de palabras capturadas.
print(f"Muestra de palabras en mayúsculas: {mayus_words}")

Total de palabras encontradas: 455
Muestra de palabras en mayúsculas: ['I', 'I', 'THAT', 'A', 'I', 'I', 'I', 'I', 'I', 'I', 'I', 'I', 'I', 'I', 'I', 'I', 'APPETIZERS', 'A', 'I', 'I', 'I', 'I', 'I', 'I', 'I', 'WILL', 'NEVER', 'EVER', 'STEP', 'FORWARD', 'IN', 'IT', 'AGAIN', 'I', 'LOVED', 'I', 'AND', 'REAL', 'I', 'I', 'I', 'I', 'BITCHES', 'I', 'I', 'I', 'I', 'NYC', 'I', 'I', 'I', 'I', 'I', 'I', 'I', 'STALE', 'I', 'I', 'DELICIOUS', 'I', 'I', 'I', 'I', 'I', 'I', 'I', 'I', 'I', 'I', 'I', 'I', 'I', 'I', 'I', 'WORST', 'EXPERIENCE', 'EVER', 'I', 'ALL', 'I', 'BARGAIN', 'I', 'TV', 'NONE', 'I', 'M', 'FREEZING', 'AYCE', 'I', 'I', 'I', 'I', 'I', 'I', 'I', 'I', 'I', 'I', 'I', 'I', 'I', 'FLAVOR', 'I', 'I', 'I', 'I', 'I', 'I', 'I', 'I', 'I', 'I', 'I', 'I', 'I', 'I', 'I', 'I', 'I', 'I', 'NEVER', 'I', 'I', 'I', 'I', 'BBQ', 'I', 'A', 'I', 'UNREAL', 'I', 'I', 'I', 'I', 'I', 'I', 'OMG', 'I', 'BETTER', 'I', 'I', 'BLAND', 'I', 'I', 'I', 'I', 'I', 'I', 'I', 'RUDE', 'INCONSIDERATE', 'MANAGEMENT', 'I', 'I', 'WIL

*   **Pregunta 4.**  

Busca e imprime los comentarios en donde todos los caracteres alfabéticos (letras) están en mayúsculas. 

Cada coincidencia encontrada debe ser todo el comentario/enunciado.

Indica cuántos resultados obtuviste.


In [50]:
# Filtramos comentarios que contienen al menos una letra mayúscula y ninguna letra minúscula.
full_mayus = [c for c in com_clean if re.search(r'[A-Z]', c) and not re.search(r'[a-z]', c)]

In [51]:
# Mostramos una muestra de los primeros 5 comentarios que cumplen con estar escritos totalmente en mayúsculas.
print(f"Muestra de comentarios en mayúsculas: {full_mayus[:5]}")

# Imprimimos el total de comentarios encontrados que coinciden con el patrón de "solo mayúsculas".
print(f"Total de comentarios detectados: {len(full_mayus)}")

Muestra de comentarios en mayúsculas: ['DELICIOUS!!', 'RUDE & INCONSIDERATE MANAGEMENT.', 'WILL NEVER EVER GO BACK AND HAVE TOLD MANY PEOPLE WHAT HAD HAPPENED.', 'TOTAL WASTE OF TIME.', 'AVOID THIS ESTABLISHMENT!']
Total de comentarios detectados: 5


*   **Pregunta 5.**  

Busca e imprime todas las palabras que tengan una vocal acentuada, del tipo á, é, í, ó, ú. 

Indica cuántos resultados obtuviste.

In [52]:
# Utilizamos re.findall para buscar palabras (\w*) que contengan al menos una vocal acentuada [áéíóúÁÉÍÓÚ] en cualquier posición.
vocal_words = re.findall(r'\b\w*[áéíóúÁÉÍÓÚ]\w*\b', " ".join(com_clean))

In [53]:
# Mostramos la lista de palabras encontradas para verificar.
print(f"Palabras con vocales acentuadas: {vocal_words}")

# Imprimimos el total de coincidencias.
print(f"Total de resultados obtenidos: {len(vocal_words)}")

Palabras con vocales acentuadas: []
Total de resultados obtenidos: 0


*   **Pregunta 6.**  

Busca e imprime todas las cantidades numéricas monetarias, enteras o con decimales, que inician con el símbolo $\$$. 

Indica cuántos resultados obtuviste.

In [54]:
# Buscamos en todo el texto unido (join) todas las coincidencias que empiecen con el símbolo $ seguido de dígitos y posibles separadores decimales.
cant_dinero = re.findall(r'\$\d+(?:[.,]\d+)*', " ".join(docs))

In [55]:
# Imprimimos el total de menciones monetarias encontradas para conocer la frecuencia de precios en el corpus.
print(f"Total de cantidades monetarias encontradas: {len(cant_dinero)}")

print(f"Cantidades monetarias encontradas: {cant_dinero}")

Total de cantidades monetarias encontradas: 8
Cantidades monetarias encontradas: ['$20', '$4.00', '$17', '$3', '$35', '$7.85', '$12', '$11.99']


*   **Pregunta 7.**  

Busca e imprime todas las palabras que sean variantes de la palabra "love", sin importar si incluyen mayúsculas o minúsculas, o la manera en que esté conjugada o alguna otra variación que se haga con dicha palabra.

Indica cuántos resultados obtuviste.

In [56]:
# Utilizamos re.findall con la bandera re.IGNORECASE para encontrar todas las palabras que comiencen con la raíz 'lov'.
# El patrón \blov\w*\b asegura que capturemos la palabra completa (como 'loved' o 'loving') independientemente de su terminación.
love_variants = re.findall(r'\blov\w*\b', " ".join(docs), re.IGNORECASE)

In [57]:
# Imprimimos la lista completa de las variantes encontradas para analizar las distintas formas de expresar afecto en las reseñas.
print(f"Variantes de 'love' encontradas: {love_variants}")

# Mostramos el conteo total de variantes encontradas.
print(f"Total de resultados obtenidos: {len(love_variants)}")

Variantes de 'love' encontradas: ['Loved', 'loved', 'Loved', 'love', 'loves', 'LOVED', 'lovers', 'loving', 'love', 'lovers', 'Love', 'loved', 'loved', 'love', 'love', 'love', 'loved', 'love', 'loved', 'Love', 'LOVED', 'love', 'lovely', 'love', 'lovely', 'love', 'lover', 'loved', 'love', 'love', 'love', 'love', 'love', 'love', 'love', 'love']
Total de resultados obtenidos: 36


*   **Pregunta 8.**  

Busca e imprime todas las palabras, variantes de "so" y "good", que tengan dos o más "o" en "so" y 3 o más "o" en good. 

Indica cuántas encontraste.


In [58]:
# Buscamos variantes de "so" con 2 o más letras "o" y de "good" con 3 o más letras "o".
so_good_variants = re.findall(r'\bso{2,}\b|\bgo{3,}d\b', " ".join(docs), re.IGNORECASE)

In [59]:
# Imprimimos la lista de las exageraciones encontradas para ver cómo los usuarios enfatizan sus reseñas.
print(f"Exageraciones encontradas: {so_good_variants}")

# Mostramos el total de coincidencias encontradas.
print(f"Total: {len(so_good_variants)}")

Exageraciones encontradas: ['Sooooo', 'soooo', 'soooooo', 'soooo']
Total: 4


*   **Pregunta 9.**  

Busca e imprime todas las palabras que tengan una longitud mayor estrictamente a 10 caracteres alfabéticos.

No se consideran los signos de puntuación o caracteres especiales en la longitud de estas cadenas, solo caracteres alfabéticos en mayúsculas o minúsculas.

Indica la cantidad de palabras encontradas.


In [60]:
palabras_largas = [p for p in re.findall(r'\b[A-Za-z]+\b', " ".join(docs)) if len(p) > 10 ]

In [61]:
# Imprimimos la lista completa de palabras largas encontradas.
print(f"Palabras de más de 10 letras: {palabras_largas}")

# Mostramos el conteo total de palabras encontradas.
print(f"Total encontrado: {len(palabras_largas)}")

Palabras de más de 10 letras: ['recommendation', 'recommended', 'overwhelmed', 'inexpensive', 'establishment', 'imaginative', 'opportunity', 'experiencing', 'underwhelming', 'relationship', 'unsatisfying', 'disappointing', 'outrageously', 'disappointing', 'expectations', 'restaurants', 'suggestions', 'disappointed', 'considering', 'Unfortunately', 'immediately', 'ingredients', 'accommodations', 'maintaining', 'Interesting', 'disrespected', 'accordingly', 'unbelievable', 'cheeseburger', 'descriptions', 'inexpensive', 'disappointed', 'Veggitarian', 'outstanding', 'recommendation', 'disappointed', 'disappointed', 'neighborhood', 'disappointed', 'corporation', 'considering', 'exceptional', 'shawarrrrrrma', 'disappointed', 'vinaigrette', 'immediately', 'unbelievably', 'replenished', 'disappointed', 'enthusiastic', 'Outstanding', 'comfortable', 'interesting', 'INCONSIDERATE', 'considering', 'transcendant', 'disappointment', 'disappointed', 'disappointed', 'overwhelmed', 'professional', 'Furt

*   **Pregunta 10.**  

Busca e imprime todas las palabras que inician con una letra mayúscula y terminan con una minúscula, pero que además no sea la primera palabra del comentario/string. 

Indica la cantidad de resultados obtenidos.

In [62]:
# Busca palabras por comentario, excluye la primera palabra de cada uno con [1:],
# y conserva solo las que inician con mayúscula y terminan con minúscula.
palabras_p10 = [
    palabra
    for comentario in docs
    for palabra in re.findall(r'\b[A-Za-zÁÉÍÓÚÑáéíóúñ]+\b', comentario)[1:]
    if re.fullmatch(r'[A-ZÁÉÍÓÚÑ][A-Za-zÁÉÍÓÚÑáéíóúñ]*[a-záéíóúñ]', palabra)
]

In [63]:
# Imprimimos la lista de palabras encontradas que cumplen con la condición de capitalización y posición.
print(f"Palabras encontradas (excluyendo la primera del comentario): {palabras_p10}")

# Mostramos el conteo total de coincidencias.
print(f"Total de resultados obtenidos: {len(palabras_p10)}")

Palabras encontradas (excluyendo la primera del comentario): ['Loved', 'May', 'Rick', 'Steve', 'Cape', 'Cod', 'Vegas', 'Burrittos', 'Blah', 'The', 'Mexican', 'Luke', 'Our', 'Hiro', 'Firehouse', 'Greek', 'Greek', 'Heart', 'Attack', 'Grill', 'Vegas', 'Dos', 'Gringos', 'Jeff', 'Excalibur', 'Very', 'Bad', 'Customer', 'Service', 'Vegas', 'Rice', 'Company', 'Pho', 'Hard', 'Rock', 'Casino', 'Buffet', 'Tigerlilly', 'Yama', 'Thai', 'Indian', 'Not', 'Vegas', 'Lox', 'Subway', 'Subway', 'Vegas', 'Vegas', 'Mandalay', 'Bay', 'Great', 'Voodoo', 'Phoenix', 'Vegas', 'Khao', 'Soi', 'Lemon', 'Joey', 'Valley', 'Phoenix', 'Magazine', 'Pho', 'Fridays', 'Tasty', 'Jamaican', 'Bisque', 'Bussell', 'Sprouts', 'Risotto', 'Filet', 'Otto', 'Yeah', 'Honestly', 'Not', 'Also', 'Vegas', 'Greek', 'Vegas', 'Veggitarian', 'Madison', 'Ironman', 'Jenni', 'Pho', 'Bachi', 'Burger', 'Pizza', 'Salads', 'They', 'Yelpers', 'Bachi', 'Service', 'English', 'Pizza', 'Hut', 'Seat', 'Gold', 'Standard', 'Thai', 'Tucson', 'Vegas', 'Chipo

*   **Pregunta 11.**  

Busca e imprime la secuencia de dos o más palabras que están separadas por un guion, "-", sin que tengan espacios en blanco entre ellas.

Por ejemplo "Go-Kart" sería válido, pero "Go  -Kart" o "Go  -  Kart" no lo serían. 

Indica la cantidad de resultados obtenidos.

In [64]:
# Buscamos secuencias de dos o más palabras alfabéticas unidas por guion, sin espacios.

palabras_compuestas = re.findall(r'\b[A-Za-z]+(?:-[A-Za-z]+)+\b', " ".join(docs))

In [65]:
# Imprimimos la lista de secuencias con guion encontradas para analizar cómo se estructuran estos términos en las reseñas.
print(f"Secuencias con guion encontradas: {palabras_compuestas}")

# Mostramos el total de resultados obtenidos.
print(f"Total: {len(palabras_compuestas)}")

Secuencias con guion encontradas: ['flat-lined', 'hands-down', 'must-stop', 'sub-par', 'Service-check', 'in-house', 'been-stepped-in-and-tracked-everywhere', 'multi-grain', 'to-go', 'non-customer', 'High-quality', 'sit-down', 'over-whelm', 'low-key', 'non-fancy', 'golden-crispy', 'over-priced', 'over-hip', 'under-services']
Total: 19


*   **Pregunta 12.**  

Busca e imprime todas las palabras que terminan en "ing" o "ed". 

Indica la cantidad de palabras que encontraste de cada una.

In [66]:
# Unimos el texto y ejecutamos ambas búsquedas en una sola línea usando asignación múltiple.
res_ing, res_ed = re.findall(r'\b\w+ing\b', " ".join(docs), re.I), re.findall(r'\b\w+ed\b', " ".join(docs), re.I)

In [67]:
print(f"Palabras terminadas en 'ing': {res_ing}")
print(f"Total 'ing': {len(res_ing)}")

print(f"\nPalabras terminadas en 'ed': {res_ed}")
print(f"Total 'ed': {len(res_ed)}")

Palabras terminadas en 'ing': ['during', 'getting', 'being', 'being', 'amazing', 'running', 'redeeming', 'getting', 'thing', 'dressing', 'refreshing', 'running', 'amazing', 'nothing', 'appalling', 'wasting', 'eating', 'going', 'Coming', 'experiencing', 'underwhelming', 'eating', 'raving', 'spring', 'unsatisfying', 'amazing', 'Everything', 'disappointing', 'dining', 'flirting', 'thing', 'coming', 'playing', 'ordering', 'arriving', 'disappointing', 'preparing', 'loving', 'liking', 'reviewing', 'venturing', 'including', 'during', 'changing', 'going', 'considering', 'coming', 'going', 'everything', 'looking', 'dressing', 'dining', 'Everything', 'amazing', 'judging', 'maintaining', 'asking', 'having', 'something', 'lacking', 'Interesting', 'preparing', 'missing', 'feeling', 'exceeding', 'inviting', 'climbing', 'waiting', 'coming', 'being', 'lacking', 'going', 'amazing', 'dealing', 'annoying', 'falling', 'sporting', 'amazing', 'providing', 'building', 'FREEZING', 'lighting', 'going', 'nothin

#**Parte 3. Proceso de limpieza.**

*   **Pregunta 13.**  

Ahora realiza un proceso de limpieza del corpus que incluya los siguientes procesos:

*   Solo se deben considerar caracteres alfabéticos. Es decir, se eliminan todos los signos de puntuación y caracteres especiales.
*   Todos los caracteres alfabéticos se transforman a minúsculas.
*   Se deben eliminar todos los espacios en blanco adicionales que se puedan encontrar en cada comentario.

Al finalizar dicho proceso de limpieza, imprime el resultado de los primeros 10 comentarios resultantes. 
   




In [68]:
# Aplicamos re.sub para mantener solo letras y espacios, pasamos a minúsculas con .lower() 
# y normalizamos espacios en blanco usando .split() seguido de .join().
docs_clean = [" ".join(re.sub(r'[^A-Za-zÁÉÍÓÚÑáéíóúñ\s]', '', c).lower().split()) for c in docs]

In [69]:
# Imprimimos los primeros 10 comentarios ya procesados para validar que la limpieza fue exitosa.
print(f"Muestra del corpus limpio:\n{docs_clean[:10]}")

# Mostramos el total de documentos para confirmar que mantenemos los 1000 registros originales.
print(f"Total de comentarios procesados: {len(docs_clean)}")

Muestra del corpus limpio:
['wow loved this place', 'crust is not good', 'not tasty and the texture was just nasty', 'stopped by during the late may bank holiday off rick steve recommendation and loved it', 'the selection on the menu was great and so were the prices', 'now i am getting angry and i want my damn pho', 'honeslty it didnt taste that fresh', 'the potatoes were like rubber and you could tell they had been made up ahead of time being kept under a warmer', 'the fries were great too', 'a great touch']
Total de comentarios procesados: 1000


*   **Pregunta 14.**  

Con el resultado de la limpieza obtenido en la pregunta anterior, realiza ahora un proceso de tokenización por palabras del corpus. 

Es decir, al final de este proceso de tokenización, debes tener como resultado una lista de listas, donde cada comentario estará tokenizado por palabras.

Al terminar calcula el total de tokens obtenido en todo el corpus. 

In [70]:
# Creamos una lista de listas donde cada sublista contiene las palabras (tokens) de un comentario.
# Usamos .split() que, por defecto, separa por cualquier espacio en blanco.
tokens = [c.split() for c in docs_clean]

In [71]:
# Imprimimos los tokens de los primeros 5 comentarios para observar la estructura de lista de listas.
print(f"Muestra de la tokenización (primeros 5): {tokens[:5]}")

# Calculamos el total de tokens generados en todo el corpus para conocer el volumen de datos.
print(f"Total de tokens en el corpus: {sum(len(t) for t in tokens)}")

Muestra de la tokenización (primeros 5): [['wow', 'loved', 'this', 'place'], ['crust', 'is', 'not', 'good'], ['not', 'tasty', 'and', 'the', 'texture', 'was', 'just', 'nasty'], ['stopped', 'by', 'during', 'the', 'late', 'may', 'bank', 'holiday', 'off', 'rick', 'steve', 'recommendation', 'and', 'loved', 'it'], ['the', 'selection', 'on', 'the', 'menu', 'was', 'great', 'and', 'so', 'were', 'the', 'prices']]
Total de tokens en el corpus: 10777


*   **Pregunta 15.**  

Finalmente, en este ejercicio definiremos nuestro conjunto de palabras "stopwords", las cuales deberás eliminar de todo el corpus. 

Recuerda que ejemplos de stopwords son artículos, adverbios, conectivos, etcétera, que tienen frecuencias de aparición muy altas en cualquier documento, pero que no brindan mucho significado en cuanto al significado de un enunciado.

Con base a la lista de stopwords que se te proporciona, realiza un proceso de limpieza eliminando todas estas palabras del corpus obtenido en el ejercicio anterior.

Obtener cuántos tokens/palabras quedan finalmente en todo el corpus.

Obtener cuántos de estos tokens/palabras son diferentes, es decir, cuántos tokens únicos tendrá lo que llamaremos más adelante nuestro vocabulario.

In [72]:
# Considera la siguiente lista como tu conjunto de stopwords:
mis_stopwords = ['i', 'me', 'my', 'myself', 'we', 'our', 'ours', 'ourselves', 'you', 'your', 'yours', 'he', 'him', 'his', 'himself', 'she', 'her', 'hers', 'herself', 'it', 'its', 'itself', 'they', 'them', 'their', 'theirs', 'themselves', 'what', 'which', 'who', 'whom', 'this', 'that', 'these', 'those', 'am', 'is', 'are', 'was', 'were', 'be', 'been', 'being', 'have', 'has', 'had', 'having', 'do', 'does', 'did', 'doing', 'a', 'an', 'the', 'and', 'but', 'if', 'or', 'because', 'as', 'until', 'while', 'of', 'at', 'by', 'for', 'with', 'about', 'against', 'between', 'into', 'through', 'during', 'before', 'after', 'above', 'below', 'to', 'from', 'up', 'down', 'in', 'out', 'on', 'over', 'under', 'again', 'further', 'then', 'once', 'here', 'there', 'when', 'where', 'why', 'how', 'all', 'any', 'both', 'each', 'few', 'more', 'most', 'other', 'some', 'such', 'only', 'own', 'same', 'so', 'than', 'too', 'very', 's', 't', 'can', 'will', 'just', 'should', 'now', 'll']

In [73]:
# Filtramos cada lista de tokens para mantener solo las palabras que NO están en nuestra lista de stopwords.
tokens_final = [[w for w in s if w not in mis_stopwords] for s in tokens]

# Generamos el vocabulario único final a partir de los tokens ya filtrados.
vocab_final = set(p for sub in tokens_final for p in sub)

In [74]:
# Imprimimos el gran total de palabras que quedaron en todo el corpus tras la limpieza.
print(f"Tokens totales finales: {sum(len(t) for t in tokens_final)}")

# Mostramos el tamaño del vocabulario final.
print(f"Tamaño del vocabulario único: {len(vocab_final)}")

Tokens totales finales: 5776
Tamaño del vocabulario único: 1941


*   **Comentarios**

Incluye finalmente tus comentarios de la actividad.

La ejecución de esta actividad permitió establecer un flujo de trabajo profesional para la normalización de datos no estructurados mediante el uso estratégico de expresiones regulares, logrando transformar un corpus de mil reseñas crudas de Yelp en un vocabulario depurado y listo para el modelado léxico. A través de la limpieza profunda, que incluyó la eliminación de ruido estructural, la estandarización a minúsculas y el filtrado de stopwords, se logró una reducción significativa de la dimensionalidad, pasando de 10,777 tokens iniciales a 5,776 palabras con alta carga semántica. La identificación de patrones específicos, como las 267 instancias de capitalización no inicial y las variantes morfológicas en pasado y gerundio, demuestra que las técnicas de Regex no solo sirven para la limpieza, sino que actúan como una herramienta de ingeniería de características capaz de capturar la naturaleza narrativa y el énfasis pragmático de los usuarios. Finalmente el proceso subraya que la calidad de los resultados en Procesamiento de Lenguaje Natural depende directamente de la rigurosidad en la normalización, donde un vocabulario final de 1,941 términos únicos representa ahora una base sólida y eficiente para tareas avanzadas de analítica y minería de texto.

##**Fin de la Actividad de la semana 2.**